rosbag2_storage/include/rosbag2_storage/bag_metadata.hpp

```diff
@@ -0,0 +1,50 @@
+// Copyright 2018, Bosch Software Innovations GmbH.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2_STORAGE__BAG_METADATA_HPP_
+#define ROSBAG2_STORAGE__BAG_METADATA_HPP_
+
+#include <chrono>
+#include <map>
+#include <string>
+#include <vector>
+#include <utility>
+
+#include "rosbag2_storage/topic_with_type.hpp"
+
+namespace rosbag2_storage
+{
+
+struct TopicMetadata
+{
+  TopicWithType topic_with_type;
+  size_t message_count;
+};
+
+struct BagMetadata
+{
+  int version = 1;
```

Comment:
What is version 1 meaning here? How is the user supposed to change this version?

Reply:
The version number is supposed to allow later changes to the metadata format (e.g. if we also want to save info about services, actions, etc at some later point). Having a version field simplifies handling old bagfiles then.
Therefore, the user must not change this version.

Reply:
The version number is supposed to allow later changes to the metadata format (e.g. if we also want to save info about services, actions, etc at some later point). Having a version field simplifies handling old bagfiles then.
Therefore, the user must not change this version.

Reply:
The version number is supposed to allow later changes to the metadata format (e.g. if we also want to save info about services, actions, etc at some later point). Having a version field simplifies handling old bagfiles then.
Therefore, the user must not change this version.

Reply:
The version number is supposed to allow later changes to the metadata format (e.g. if we also want to save info about services, actions, etc at some later point). Having a version field simplifies handling old bagfiles then.
Therefore, the user must not change this version.

Reply:
The version number is supposed to allow later changes to the metadata format (e.g. if we also want to save info about services, actions, etc at some later point). Having a version field simplifies handling old bagfiles then.
Therefore, the user must not change this version.

Reply:
This version field was introduced to allow easy transition to new formats if necessary (e.g. we want to add information about service calls, parameters, etc.). Having a version field then simplifies backwards compatibility with old yaml files.

Therefore, it should never be changed by the user.

Reply:
This version field was introduced to allow easy transition to new formats if necessary (e.g. we want to add information about service calls, parameters, etc.). Having a version field then simplifies backwards compatibility with old yaml files.

Therefore, it should never be changed by the user.

Reply:
This version field was introduced to allow easy transition to new formats if necessary (e.g. we want to add information about service calls, parameters, etc.). Having a version field then simplifies backwards compatibility with old yaml files.

Therefore, it should never be changed by the user.

Reply:
This version field was introduced to allow easy transition to new formats if necessary (e.g. we want to add information about service calls, parameters, etc.). Having a version field then simplifies backwards compatibility with old yaml files.

Therefore, it should never be changed by the user.

Reply:
where is this field currently used in the code? Can we extract that info somehow from incoming messages?

Reply:
It is not used yet and its sole purpose is to distinguish the yaml format in case of changes. It must be independent of the content so a md5 sum (or something similar) is not useful.
An example use would be the case of a format change (after the first release). For that case rosbag can distinguish the to formats and provide an automatic conversion (or at least read also the old format correctly).